packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js

```diff
@@ -1,142 +1,184 @@
-	return createBlock(
-		'core/navigation-link',
-		{
-			label: menuItem.title.rendered,
-			url: menuItem.url,
-		},
-		innerBlocks
-	);
-}
-
-function createMenuItemAttributesFromBlock( block ) {
-	return {
-		title: block.attributes.label,
-		url: block.attributes.url,
-	};
-}
+/**
+ * Internal dependencies
+ */
+import batchSave from './batch-save';
+import useDebouncedValue from './use-debounced-value';
+import PromiseQueue from './promise-queue';
 
 export default function useNavigationBlocks( menuId ) {
+	const [ deletedMenuItemsIds, setDeletedMenuItemsIds ] = useState( [] );
+
 	// menuItems is an array of menu item objects.
+	const query = { menus: menuId, per_page: -1 };
 	const menuItems = useSelect(
 		( select ) =>
-			select( 'core' ).getMenuItems( { menus: menuId, per_page: -1 } ),
+			select( 'core' )
+				.getMenuItems( query )
+				?.filter( ( { id } ) => ! deletedMenuItemsIds.includes( id ) ),
 		[ menuId ]
 	);
 
-	const { saveMenuItem } = useDispatch( 'core' );
-	const { createSuccessNotice } = useDispatch( 'core/notices' );
+	// Data model
 	const [ blocks, setBlocks ] = useState( [] );
-
 	const menuItemsRef = useRef( {} );
 
+	// Refresh our model whenever menuItems change
 	useEffect( () => {
-		if ( ! menuItems ) {
-			return;
+		if ( menuItems ) {
+			const [
+				navigationBlock,
+				clientIdToMenuItemMapping,
+			] = menuItemsToNavigationBlock( menuItems );
+			setBlocks( [ navigationBlock ] );
+			menuItemsRef.current = clientIdToMenuItemMapping;
 		}
+	}, [ menuItems ] );
 
-		const itemsByParentID = groupBy( menuItems, 'parent' );
-
-		menuItemsRef.current = {};
-
-		const createMenuItemBlocks = ( items ) => {
-			const innerBlocks = [];
-			if ( ! items ) {
-				return;
-			}
-			for ( const item of items ) {
-				let menuItemInnerBlocks = [];
-				if ( itemsByParentID[ item.id ]?.length ) {
-					menuItemInnerBlocks = createMenuItemBlocks(
-						itemsByParentID[ item.id ]
-					);
+	// When a new block is added, let's create a draft menuItem for it.
+	// The batch save endpoint expects all the menu items to have a valid id already.
+	// PromiseQueue is used in order to
+	// 1) limit the amount of requests processed at the same time
+	// 2) save the menu only after all requests are finalized
+	const debouncedBlocks = useDebouncedValue( blocks, 800 );
+	const promiseQueueRef = useRef( new PromiseQueue() );
+	const processedBlocksIds = useRef( [] );
+	useEffect(
+		function() {
+			for ( const clientId of getAllClientIds( debouncedBlocks ) ) {
+				// Already has a related menu item
+				if ( clientId in menuItemsRef.current ) {
+					continue;
 				}
-				const block = createBlockFromMenuItem(
-					item,
-					menuItemInnerBlocks
+				// Already being processed
+				if ( processedBlocksIds.current.includes( clientId ) ) {
+					continue;
+				}
+				processedBlocksIds.current.push( clientId );
+				promiseQueueRef.current.enqueue( () =>
+					createDraftMenuItem( clientId ).then(
```

Comment:
I was under the impression that we'd save draft menu items when they're added in the navigation block.

Reply:
I'm not sure I follow - this code does exactly that.